components/ai_chat/resources/page/state/ai_chat_context.tsx

```diff
@@ -77,63 +66,10 @@ export function AIChatContextProvider(props: React.PropsWithChildren<Props>) {
   }
 
   React.useEffect(() => {
-    const { Service, Observer } = getAPI()
-    async function initialize() {
-      const [
-        { conversations: visibleConversations },
-        { actionList: allActions },
-        { canShow: canShowPremiumPrompt }
-      ] = await Promise.all([
-        Service.getVisibleConversations(),
-        Service.getActionMenuList(),
-        Service.getCanShowPremiumPrompt()
-      ])
-      setPartialContext({
-        visibleConversations,
-        allActions,
-        canShowPremiumPrompt
-      })
-    }
-
-    async function updateCurrentPremiumStatus() {
-      const { status } = await getAPI().Service.getPremiumStatus()
-      setPartialContext({
-        isPremiumStatusFetching: false,
-        isPremiumUser: (status !== undefined && status !== mojom.PremiumStatus.Inactive),
-        isPremiumUserDisconnected: status === mojom.PremiumStatus.ActiveDisconnected
-      })
-    }
-
-    initialize()
-    updateCurrentPremiumStatus()
-
-    if (context.isHistoryEnabled) {
-      Observer.onConversationListChanged.addListener(
-        (conversations: mojom.Conversation[]) => {
-          setPartialContext({
-            visibleConversations: conversations
-          })
-        }
-      )
-    }
+    setPartialContext(getAPI().UIState)
 
-    Observer.onAgreementAccepted.addListener(() =>
-      setPartialContext({
-        hasAcceptedAgreement: true
-      })
-    )
-
-    // Since there is no server-side event for premium status changing,
-    // we should check often. And since purchase or login is performed in
-    // a separate WebContents, we can check when focus is returned here.
-    window.addEventListener('focus', () => {
-      updateCurrentPremiumStatus()
-    })
-
-    document.addEventListener('visibilitychange', (e) => {
-      if (document.visibilityState === 'visible') {
-        updateCurrentPremiumStatus()
-      }
+    getAPI().addUIStateChangeListener((e) => {
```

Comment:
looking at this, does using the `CustomEvent` API buy as anything except needing to access the data through `.detail`? We don't need it to bubble or anything (and it doesn't)